### PR TITLE
API code review 

### PR DESCRIPTION
**Pull request by accelerated**
**Description of changes**
* Code clean-up and readability improvements based on internal code-review. No functionality changes were introduced except for small API signature changes below.

**ABI/API breaking changes**
* `ProducerConfiguration` constructor now takes a full topic object as parameter instead of just the topic name string. This is to make it identical to `ConsumerConfiguration`. 
* The `ConsumerConfiguration` constructor overload which didn't take a `ReceiverCallback` have been removed as well as `ConsumerConfiguration::setReceiverCallback()`. This is to eliminate the confusion related to having two setters (one in the constructor and another post-construction).
* `ConsumerManager::preprocess()` has been renamed `ConsumerManager::enablePreprocessing()`/`ConsumerManager::disablePreprocessing()`
* API taking an empty topic string to imply _all topics_ have now been split into two overloads (one taking a topic and another no taking any arguments). This change is API backwards compatible.
* `ConsumerManager::commit()` functions as well as `OffsetManager::saveOffset()` functions have the `forceSync` boolean parameter converted to an _enum_ for better readability. 
* `HeaderPack::get()` has its default 2nd argument changed from `-1` to `1`.

### CODE REVIEW

File: corokafka/corokafka_header_pack.cpp [hunk 1]
```diff
@@ -55,38 +56,46 @@ HeaderPack& HeaderPack::pop_back() {
...
+        if (StringEqualCompare()(header.first, name)) {
+            return (relativePosition == 0) || (pos++ == relativePosition);
+        }
+        return false;
     }), _headers.end());
 }
 
-size_t HeaderPack::size() const {
+void HeaderPack::erase(const HeaderPack::ListType::iterator& it) {
+    _headers.erase(it);
```
Comment (safaruqi) [inline corokafka/corokafka_header_pack.cpp:70]: can we check if this iterator is valid here. at least check it is not end() or past it

File: corokafka/corokafka_header_pack.cpp [hunk 2]
```diff
@@ -55,38 +56,46 @@ HeaderPack& HeaderPack::pop_back() {
...
 }
 
-void HeaderPack::erase(const std::string& name) {
-    _headers.erase(std::remove_if(_headers.begin(), _headers.end(), [&name](const HeaderNode& entry)->bool {
-        return StringEqualCompare()(entry.first, name);
+void HeaderPack::erase(const std::string& name, size_t relativePosition) {
+    size_t pos = 1;
+    _headers.erase(std::remove_if(_headers.begin(), _headers.end(), [&](const HeaderNode& header)->bool {
+        if (StringEqualCompare()(header.first, name)) {
+            return (relativePosition == 0) || (pos++ == relativePosition);
```
Comment (safaruqi) [inline corokafka/corokafka_header_pack.cpp:63]: == should be >= if I understand correctly what is being done here.

Comment (accelerated) [reply corokafka/corokafka_header_pack.cpp:63]: We want to stop when we reach the Nth header with this name. The logic is ok. If 0 is passed, then it's the first occurrence.

File: corokafka/impl/corokafka_consumer_manager_impl.cpp [hunk 3]
```diff
@@ -428,98 +446,86 @@ void ConsumerManagerImpl::subscribe(const std::string& topic,
...
-    else {
-        ConsumerTopicEntry& topicEntry = findConsumer(topic)->second;
-        if (!topicEntry._isSubscribed) {
-            subscribe(topic, topicEntry, partitionList);
-        }
-        else {
-            throw ConsumerException(topic, "Consumer already subscribed");
-        }
+    for (auto&& consumer : _consumers) {
+        unsubscribe(consumer.first);
```
[the remark this answers is not in the document]
Comment (accelerated) [reply corokafka/impl/corokafka_consumer_manager_impl.cpp:452]: Ok.

**Pull request merged**